src/mpid/ch4/netmod/ofi/ofi_am_impl.c

```diff
@@ -22,73 +23,78 @@ void MPIDI_OFI_deferred_am_isend_dequeue(MPIDI_OFI_deferred_am_isend_req_t * dre
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_DEFERRED_AM_ISEND_ISSUE);
-
-    MPIDI_Datatype_get_info(dreq->count, dreq->datatype, dt_contig, data_sz, dt_ptr, dt_true_lb);
-    need_packing = dt_contig ? false : true;
-    send_buf = (char *) dreq->buf + dt_true_lb;
+    MPI_Aint packed_size;
+    MPI_Aint dt_true_lb;
+    MPIR_Datatype *dt_ptr;
+    MPIDI_OFI_am_send_request_t *send_req;
 
-    do_eager =
-        (MPIDI_OFI_AMREQUEST_HDR(dreq->sreq, am_hdr_sz) + dreq->data_sz
-         + sizeof(MPIDI_OFI_am_header_t) <= MPIDI_OFI_DEFAULT_SHORT_SEND_SIZE);
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_ISSUE_DEFERRED_AM_ISEND_EAGER);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_ISSUE_DEFERRED_AM_ISEND_EAGER);
 
-    MPL_pointer_attr_t attr;
-    MPIR_GPU_query_pointer_attr(dreq->buf, &attr);
-    if (attr.type == MPL_GPU_POINTER_DEV && !MPIDI_OFI_ENABLE_HMEM) {
-        /* Force packing of GPU buffer in host memory */
-        need_packing = true;
-    }
-
-    if (need_packing) {
+    if (dreq->need_packing) {
         /* FIXME: currently we always do packing, also for high density types. However,
          * we should not do packing unless needed. Also, for large low-density types
          * we should not allocate the entire buffer and do the packing at once. */
-        /* TODO: (1) Skip packing for high-density datatypes;
-         *       (2) Pipeline allocation for low-density datatypes; */
-        if (do_eager) {
-            mpi_errno = MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.pack_buf_pool,
-                                                           (void **) &send_buf);
-            if (send_buf == NULL) {
-                /* no buffer available, suppress error and exit */
-                mpi_errno = MPI_SUCCESS;
-                goto fn_exit;
-            }
-            MPIR_ERR_CHECK(mpi_errno);
-        } else {
-            /* only for large message RDMA read */
-            MPL_gpu_malloc_host((void **) &send_buf, data_sz);
+        /* TODO: (1) Skip packing for high-density datatypes; */
+        mpi_errno = MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.pack_buf_pool,
+                                                       (void **) &send_buf);
+        if (send_buf == NULL) {
+            /* no buffer available, suppress error and exit */
+            mpi_errno = MPI_SUCCESS;
+            goto fn_exit;
         }
+        MPIR_ERR_CHECK(mpi_errno);
         mpi_errno = MPIR_Typerep_pack(dreq->buf, dreq->count, dreq->datatype, 0, send_buf,
-                                      dreq->data_sz, &last);
+                                      dreq->data_sz, &packed_size);
         MPIR_ERR_CHECK(mpi_errno);
+        MPIR_Assert(dreq->data_sz == packed_size);
 
-        MPIDI_OFI_AMREQUEST_HDR(dreq->sreq, pack_buffer) = send_buf;
+        MPIDI_OFI_am_create_send_request(dreq->sreq, dreq->am_hdr, dreq->am_hdr_sz, &send_req);
```

Comment:
We could directly send the `sreq` object as we did before, right? if we have enqueued the `sreq` as I commented earlier, here we can directly send from it.

I understand for pipeline send, each segment will need new buffer for each segment -- or do we? Since we always send segment one after another in a critical section, we could simply increment the segment index or update the offset, and directly send from the `sreq` too, right?